pilot/pkg/bootstrap/server.go

```diff
@@ -1191,3 +1204,56 @@ func (s *Server) waitForCacheSync(stop <-chan struct{}) bool {
 
 	return true
 }
+
+func (s *Server) initCertController(args *PilotArgs) error {
+	var err error
+	var svcNames []string
+	var svcNamespaces []string
+	if !s.mesh.K8SCertificateSetting.Enabled {
+		log.Info("k8s certificate provision is not enabled")
+		return nil
+	}
+
+	k8sClient := s.kubeClient
+	if len(s.mesh.K8SCertificateSetting.PilotService) > 0 {
+		// Generate a key and certificate for Pilot and save it
+		// a directory.
+		pilotSvcNs := strings.Split(s.mesh.K8SCertificateSetting.PilotService, ".")
+		if len(pilotSvcNs) != 2 {
+			log.Error("pilot service must have a service name and service namespace, delimited by a '.'")
+			return nil
+		}
+		// TO-DO (lei-tang):
+		// - Make a directory at s.mesh.K8SCertificateSetting.PilotCertificatePath
+		// - Generate Pilot certificate by calling
+		// chiron.GenKeyCertK8sCA(k8sClient.CertificatesV1beta1(), pilotSvcNs[0] + ".secret",
+		// pilotSvcNs[1], pilotSvcNs[0], s.mesh.K8SCertificateSetting.CaBundleFile).
+		// - Save the key.pem, cert-chain.pem, and root.pem to the directory.
+	}
+
+	// Provision and manage the certificates for non-Pilot services.
+	// If services are empty, the certificate controller will do nothing.
+	for _, svc := range s.mesh.K8SCertificateSetting.Services {
+		s := strings.Split(svc, ".")
+		if len(s) != 2 {
```

Comment:
ditto

Reply:
The format of s.mesh.K8SCertificateSetting.Services is described in istio/api#1105:
  // Other services that will be provisioned with certificates.
  // E.g., istio-galley.istio-system, istio-sidecar-injector.istio-system, and etc.
  // The format is service-name.service-namespace, with '.' as the delimiter.
  repeated string services = 5;